File: Sources/HoverPocket/PocketApps/PocketAppCapabilityMigrator.swift
import Foundation

struct PocketAppCapabilityMigrationReceipt: Equatable, Sendable {
    let packageID: String
    let sourceVersion: String
    let targetVersion: String
    let sourcePackageDigest: String
    let targetPackageDigest: String
    let migrationIDs: [String]
    let replacementCounts: [String: Int]
    let stateSchemaDigest: String
    let userDataStore: String
}

enum PocketAppCapabilityMigrationError: Error, Equatable, Sendable {
    case invalid(String)
}

struct PocketAppCapabilityMigrator {
    private let runtime: PocketAppPackageRuntime
    private let catalog: PocketCapabilityCompatibilityCatalog

    init(
        descriptors: [PocketCapabilityDescriptor] = PocketCapabilityDescriptors.builtIn,
        catalog: PocketCapabilityCompatibilityCatalog = .builtIn
    ) {
        self.runtime = PocketAppPackageRuntime(
            descriptors: descriptors,
            compatibilityCatalog: catalog
        )
        self.catalog = catalog
    }

    func migrate(
        sourceDirectory: URL,
        destinationDirectory: URL,
        targetVersion: String
    ) throws -> PocketAppCapabilityMigrationReceipt {
        guard sourceDirectory.standardizedFileURL != destinationDirectory.standardizedFileURL else {
            throw PocketAppCapabilityMigrationError.invalid("destination")
        }
        let sourceSnapshot = try PocketAppFileSnapshot.capture(directory: sourceDirectory)
        let sourcePackage = try runtime.loadMigrationSource(snapshot: sourceSnapshot)
        guard !sourcePackage.compatibilityIssues.isEmpty,
              Self.validVersion(targetVersion),
              Self.compareVersions(sourcePackage.manifest.version, targetVersion) == .orderedAscending else {
            throw PocketAppCapabilityMigrationError.invalid("version")
        }

        let migrations = try sourcePackage.compatibilityIssues.map { try catalog.migration(for: $0.key) }
        let rewritten = try rewrite(
            sourceSnapshot,
            package: sourcePackage,
            targetVersion: targetVersion,
            migrations: migrations,
            destinationDirectory: destinationDirectory
        )
        let validatedTarget = try runtime.load(snapshot: rewritten.snapshot)
        guard validatedTarget.manifest.id == sourcePackage.manifest.id,
              validatedTarget.manifest.version == targetVersion,
              validatedTarget.manifest.stateStore == sourcePackage.manifest.stateStore,
              validatedTarget.stateSchemaDigest == sourcePackage.stateSchemaDigest,
              validatedTarget.compatibilityIssues.isEmpty else {
            throw PocketAppCapabilityMigrationError.invalid("target_readback")
        }

        try rewritten.snapshot.materialize(at: destinationDirectory)
        let materialized = try runtime.load(directory: destinationDirectory)
        guard materialized.manifestDigest == validatedTarget.manifestDigest,
              materialized.stateSchemaDigest == sourcePackage.stateSchemaDigest,
              materialized.manifest.stateStore == sourcePackage.manifest.stateStore else {
            throw PocketAppCapabilityMigrationError.invalid("materialized_readback")
        }

        return PocketAppCapabilityMigrationReceipt(
            packageID: sourcePackage.manifest.id,
            sourceVersion: sourcePackage.manifest.version,
            targetVersion: targetVersion,
            sourcePackageDigest: sourcePackage.manifestDigest,
            targetPackageDigest: materialized.manifestDigest,
            migrationIDs: migrations.map(\.id).sorted(),
            replacementCounts: rewritten.counts,
            stateSchemaDigest: materialized.stateSchemaDigest,
            userDataStore: materialized.manifest.stateStore
        )
    }

    func rewriteForVerification(
        snapshot: PocketAppFileSnapshot,
        package: PocketAppPackage,
        targetVersion: String,
        migrations: [PocketCapabilityReferenceMigration],
        destinationDirectory: URL
    ) throws -> (snapshot: PocketAppFileSnapshot, counts: [String: Int]) {
        try rewrite(
            snapshot,
            package: package,
            targetVersion: targetVersion,
            migrations: migrations,
            destinationDirectory: destinationDirectory
        )
    }

    private func rewrite(
        _ snapshot: PocketAppFileSnapshot,
        package: PocketAppPackage,
        targetVersion: String,
        migrations: [PocketCapabilityReferenceMigration],
        destinationDirectory: URL
    ) throws -> (snapshot: PocketAppFileSnapshot, counts: [String: Int]) {
        var files = snapshot.files
        var counts = Dictionary(uniqueKeysWithValues: migrations.map { ($0.id, 0) })
        let bySource = Dictionary(uniqueKeysWithValues: migrations.map { ($0.source, $0) })
        let referenceMap = Dictionary(uniqueKeysWithValues: migrations.map {
            (Self.reference($0.source), Self.reference($0.target))
        })

        guard let manifestData = files["manifest.json"],
              var manifest = try Self.object(manifestData),
              var requested = manifest["requestedCapabilities"] as? [[String: Any]] else {
            throw PocketAppCapabilityMigrationError.invalid("manifest")
        }
        manifest["version"] = targetVersion
        for index in requested.indices {
            guard let id = requested[index]["id"] as? String,
                  let version = Self.integer(requested[index]["version"]) else {
                throw PocketAppCapabilityMigrationError.invalid("manifest_capability")
            }
            let key = PocketCapabilityKey(id: id, version: version)
            guard let migration = bySource[key] else { continue }
            requested[index]["id"] = migration.target.id
            requested[index]["version"] = migration.target.version
            counts[migration.id, default: 0] += 1
        }
        manifest["requestedCapabilities"] = requested
        files["manifest.json"] = try Self.canonicalJSON(manifest)

        for path in package.manifest.workflows.values.sorted() {
            guard let data = files[path], var workflow = try Self.object(data),
                  var steps = workflow["steps"] as? [[String: Any]] else {
                throw PocketAppCapabilityMigrationError.invalid("workflow")
            }
            for index in steps.indices {
                guard let reference = steps[index]["use"] as? String else {
                    throw PocketAppCapabilityMigrationError.invalid("workflow_reference")
                }
                if let target = referenceMap[reference],
                   let migration = migrations.first(where: { Self.reference($0.source) == reference }) {
                    steps[index]["use"] = target
                    counts[migration.id, default: 0] += 1
                }
            }
            workflow["steps"] = steps
            files[path] = try Self.canonicalJSON(workflow)
        }

        for path in package.manifest.surfaces.values.sorted() {
            guard let data = files[path], var surface = try Self.object(data) else {
                throw PocketAppCapabilityMigrationError.invalid("surface")
            }
            try Self.rewriteQueries(&surface, referenceMap: referenceMap, migrations: migrations, counts: &counts)
            files[path] = try Self.canonicalJSON(surface)
        }

        guard counts.values.allSatisfy({ $0 > 0 }),
              files[package.manifest.stateSchemaPath] == snapshot.files[package.manifest.stateSchemaPath] else {
            throw PocketAppCapabilityMigrationError.invalid("migration_coverage")
        }
        return (
            PocketAppFileSnapshot(
                rootDirectory: destinationDirectory.standardizedFileURL,
                files: files,
                identities: [:]
            ),
            counts
        )
    }

    private static func rewriteQueries(
        _ object: inout [String: Any],
        referenceMap: [String: String],
        migrations: [PocketCapabilityReferenceMigration],
        counts: inout [String: Int]
    ) throws {
        for key in object.keys.sorted() {
            if key == "query", let reference = object[key] as? String,
               let target = referenceMap[reference],
               let migration = migrations.first(where: { Self.reference($0.source) == reference }) {
                object[key] = target
                counts[migration.id, default: 0] += 1
            } else if var child = object[key] as? [String: Any] {
                try rewriteQueries(&child, referenceMap: referenceMap, migrations: migrations, counts: &counts)
                object[key] = child
            } else if var array = object[key] as? [Any] {
                for index in array.indices where array[index] is [String: Any] {
                    var child = array[index] as! [String: Any]
                    try rewriteQueries(&child, referenceMap: referenceMap, migrations: migrations, counts: &counts)
                    array[index] = child
                }
                object[key] = array
            }
        }
    }

    private static func object(_ data: Data) throws -> [String: Any]? {
        try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
    }

    private static func canonicalJSON(_ object: [String: Any]) throws -> Data {
        var data = try JSONSerialization.data(
            withJSONObject: object,
            options: [.sortedKeys, .withoutEscapingSlashes]
        )
        data.append(0x0A)
        return data
    }

    private static func integer(_ value: Any?) -> Int? {
        guard let number = value as? NSNumber,
              number.doubleValue.isFinite,
              number.doubleValue.rounded() == number.doubleValue else { return nil }
        return number.intValue
    }

    private static func reference(_ key: PocketCapabilityKey) -> String {
        "\(key.id)@\(key.version)"
    }

    private static func validVersion(_ value: String) -> Bool {
        value.range(
            of: "^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(?:-[0-9A-Za-z.-]+)?$",
            options: .regularExpression
        ) != nil
    }

    private static func compareVersions(_ lhs: String, _ rhs: String) -> ComparisonResult {
        let left = lhs.split(separator: "-", maxSplits: 1)[0].split(separator: ".").compactMap { Int($0) }
        let right = rhs.split(separator: "-", maxSplits: 1)[0].split(separator: ".").compactMap { Int($0) }
        for (a, b) in zip(left, right) where a != b {
            return a < b ? .orderedAscending : .orderedDescending
        }
        if lhs == rhs { return .orderedSame }
        return lhs.contains("-") && !rhs.contains("-") ? .orderedAscending : .orderedDescending
    }
}

File: windows/src/HoverPocket.Shell/PocketApps/PocketAppCapabilityMigrator.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using HoverPocket.Shell.Capabilities;

namespace HoverPocket.Shell.PocketApps;

internal sealed record PocketAppCapabilityMigrationReceipt(
    string PackageId,
    string SourceVersion,
    string TargetVersion,
    string SourcePackageDigest,
    string TargetPackageDigest,
    IReadOnlyList<string> MigrationIds,
    IReadOnlyDictionary<string, int> ReplacementCounts,
    string StateSchemaDigest,
    string UserDataStore);

internal sealed class PocketAppCapabilityMigrationException(string code) : Exception(code)
{
    internal string Code { get; } = code;
}

internal sealed class PocketAppCapabilityMigrator
{
    private readonly PocketAppPackageRuntime _runtime;
    private readonly PocketCapabilityCompatibilityCatalog _catalog;

    internal PocketAppCapabilityMigrator(
        IEnumerable<PocketCapabilityDescriptor>? descriptors = null,
        PocketCapabilityCompatibilityCatalog? catalog = null)
    {
        _catalog = catalog ?? PocketCapabilityCompatibilityCatalog.BuiltIn;
        _runtime = new PocketAppPackageRuntime(descriptors, _catalog);
    }

    internal PocketAppCapabilityMigrationReceipt Migrate(
        string sourceDirectory,
        string destinationDirectory,
        string targetVersion)
    {
        if (Path.GetFullPath(sourceDirectory).Equals(
                Path.GetFullPath(destinationDirectory),
                StringComparison.OrdinalIgnoreCase))
        {
            throw Failure("destination");
        }
        var sourceSnapshot = PocketAppFileSnapshot.Capture(sourceDirectory);
        var sourcePackage = _runtime.LoadMigrationSource(sourceSnapshot);
        if (sourcePackage.CompatibilityIssues.Count == 0
            || !ValidVersion(targetVersion)
            || CompareVersions(sourcePackage.Manifest.Version, targetVersion) >= 0)
        {
            throw Failure("version");
        }

        var migrations = sourcePackage.CompatibilityIssues
            .Select(issue => _catalog.Migration(issue.Key))
            .ToArray();
        var rewritten = Rewrite(
            sourceSnapshot,
            sourcePackage,
            targetVersion,
            migrations,
            destinationDirectory);
        var validatedTarget = _runtime.Load(rewritten.Snapshot);
        if (validatedTarget.Manifest.Id != sourcePackage.Manifest.Id
            || validatedTarget.Manifest.Version != targetVersion
            || validatedTarget.Manifest.StateStore != sourcePackage.Manifest.StateStore
            || validatedTarget.StateSchemaDigest != sourcePackage.StateSchemaDigest
            || validatedTarget.CompatibilityIssues.Count != 0)
        {
            throw Failure("target_readback");
        }

        rewritten.Snapshot.Materialize(destinationDirectory);
        var materialized = _runtime.Load(destinationDirectory);
        if (materialized.ManifestDigest != validatedTarget.ManifestDigest
            || materialized.StateSchemaDigest != sourcePackage.StateSchemaDigest
            || materialized.Manifest.StateStore != sourcePackage.Manifest.StateStore)
        {
            throw Failure("materialized_readback");
        }

        return new PocketAppCapabilityMigrationReceipt(
            sourcePackage.Manifest.Id,
            sourcePackage.Manifest.Version,
            targetVersion,
            sourcePackage.ManifestDigest,
            materialized.ManifestDigest,
            migrations.Select(item => item.Id).Order(StringComparer.Ordinal).ToArray(),
            rewritten.Counts,
            materialized.StateSchemaDigest,
            materialized.Manifest.StateStore);
    }

    internal (PocketAppFileSnapshot Snapshot, IReadOnlyDictionary<string, int> Counts) RewriteForVerification(
        PocketAppFileSnapshot snapshot,
        PocketAppPackage package,
        string targetVersion,
        IReadOnlyList<PocketCapabilityReferenceMigration> migrations,
        string destinationDirectory) =>
        Rewrite(snapshot, package, targetVersion, migrations, destinationDirectory);

    private static (PocketAppFileSnapshot Snapshot, IReadOnlyDictionary<string, int> Counts) Rewrite(
        PocketAppFileSnapshot snapshot,
        PocketAppPackage package,
        string targetVersion,
        IReadOnlyList<PocketCapabilityReferenceMigration> migrations,
        string destinationDirectory)
    {
        var files = snapshot.Files.ToDictionary(
            item => item.Key,
            item => item.Value.ToArray(),
            StringComparer.Ordinal);
        var counts = migrations.ToDictionary(item => item.Id, _ => 0, StringComparer.Ordinal);
        var bySource = migrations.ToDictionary(item => item.Source);
        var referenceMap = migrations.ToDictionary(
            item => Reference(item.Source),
            item => Reference(item.Target),
            StringComparer.Ordinal);

        var manifest = Object(files["manifest.json"], "manifest");
        manifest["version"] = targetVersion;
        var requested = manifest["requestedCapabilities"]?.AsArray() ?? throw Failure("manifest_capability");
        foreach (var item in requested)
        {
            var request = item?.AsObject() ?? throw Failure("manifest_capability");
            var id = request["id"]?.GetValue<string>() ?? throw Failure("manifest_capability");
            var version = request["version"]?.GetValue<int>() ?? throw Failure("manifest_capability");
            var key = new PocketCapabilityKey(id, version);
            if (!bySource.TryGetValue(key, out var migration)) { continue; }
            request["id"] = migration.Target.Id;
            request["version"] = migration.Target.Version;
            counts[migration.Id]++;
        }
        files["manifest.json"] = CanonicalJson(manifest);

        foreach (var path in package.Manifest.Workflows.Values.Order(StringComparer.Ordinal))
        {
            var workflow = Object(files[path], "workflow");
            var steps = workflow["steps"]?.AsArray() ?? throw Failure("workflow");
            foreach (var item in steps)
            {
                var step = item?.AsObject() ?? throw Failure("workflow_reference");
                var reference = step["use"]?.GetValue<string>() ?? throw Failure("workflow_reference");
                if (!referenceMap.TryGetValue(reference, out var target)) { continue; }
                var migration = migrations.Single(value => Reference(value.Source) == reference);
                step["use"] = target;
                counts[migration.Id]++;
            }
            files[path] = CanonicalJson(workflow);
        }

        foreach (var path in package.Manifest.Surfaces.Values.Order(StringComparer.Ordinal))
        {
            var surface = Object(files[path], "surface");
            RewriteQueries(surface, referenceMap, migrations, counts);
            files[path] = CanonicalJson(surface);
        }

        if (counts.Values.Any(value => value == 0)
            || !files[package.Manifest.StateSchemaPath].AsSpan().SequenceEqual(snapshot.Files[package.Manifest.StateSchemaPath]))
        {
            throw Failure("migration_coverage");
        }
        return (
            new PocketAppFileSnapshot(
                Path.GetFullPath(destinationDirectory),
                files,
                new Dictionary<string, PocketAppFileIdentity>(StringComparer.Ordinal)),
            counts);
    }

    private static void RewriteQueries(
        JsonNode? node,
        IReadOnlyDictionary<string, string> referenceMap,
        IReadOnlyList<PocketCapabilityReferenceMigration> migrations,
        IDictionary<string, int> counts)
    {
        if (node is JsonObject obj)
        {
            foreach (var property in obj.ToArray())
            {
                if (property.Key == "query"
                    && property.Value is JsonValue value
                    && value.TryGetValue<string>(out var reference)
                    && referenceMap.TryGetValue(reference, out var target))
                {
                    var migration = migrations.Single(item => Reference(item.Source) == reference);
                    obj[property.Key] = target;
                    counts[migration.Id]++;
                }
                else
                {
                    RewriteQueries(property.Value, referenceMap, migrations, counts);
                }
            }
        }
        else if (node is JsonArray array)
        {
            foreach (var item in array)
            {
                RewriteQueries(item, referenceMap, migrations, counts);
            }
        }
    }

    private static JsonObject Object(byte[] data, string code)
    {
        try
        {
            return JsonNode.Parse(data, new JsonNodeOptions { PropertyNameCaseInsensitive = false }, new JsonDocumentOptions
            {
                AllowTrailingCommas = false,
                CommentHandling = JsonCommentHandling.Disallow,
                MaxDepth = 16
            })?.AsObject() ?? throw Failure(code);
        }
        catch (PocketAppCapabilityMigrationException)
        {
            throw;
        }
        catch (Exception error) when (error is JsonException or InvalidOperationException)
        {
            throw Failure(code);
        }
    }

    private static byte[] CanonicalJson(JsonObject value)
    {
        var element = JsonSerializer.SerializeToElement(value);
        using var stream = new MemoryStream();
        using (var writer = new Utf8JsonWriter(stream))
        {
            CapabilityCanonicalJson.WriteElement(writer, element);
        }
        stream.WriteByte((byte)'\n');
        return stream.ToArray();
    }

    private static string Reference(PocketCapabilityKey key) => $"{key.Id}@{key.Version}";

    private static bool ValidVersion(string value) =>
        System.Text.RegularExpressions.Regex.IsMatch(
            value,
            "^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(?:-[0-9A-Za-z.-]+)?$",
            System.Text.RegularExpressions.RegexOptions.CultureInvariant);

    private static int CompareVersions(string left, string right)
    {
        var leftCore = Version.Parse(left.Split('-', 2)[0]);
        var rightCore = Version.Parse(right.Split('-', 2)[0]);
        var core = leftCore.CompareTo(rightCore);
        if (core != 0) { return core; }
        if (left == right) { return 0; }
        return left.Contains('-') && !right.Contains('-') ? -1 : 1;
    }

    private static PocketAppCapabilityMigrationException Failure(string code) => new(code);
}
